src/zenml/orchestrators/AGENTS.md

```diff
@@ -0,0 +1,178 @@
+
+**Size constraints:** Limited to ~250 characters due to MySQL database column limit, but any ID from orchestration backends should fit.
+
+### Dynamic Pipeline Case
+
+**Key differences:**
+- There's always one initial container that gets spun up first: the "orchestration container"
+- This container creates the pipeline run
+- The ID only needs to be unique when running **inside the orchestration container**
+- Does NOT need to be unique across all step containers that get spun up later
+
+**What to use:**
+- **Kubernetes:** The pod name (`socket.gethostname()`) — it's unique per container
+- **SageMaker:** The job ID of the orchestration container
+
+### Orchestration Exception: Kubernetes
+
+Kubernetes is special because it **does** spin up an orchestration container first, even for static pipelines. This means:
+- The Kubernetes orchestrator can use the pod name as the run ID
+- It first checks for `ENV_ZENML_KUBERNETES_RUN_ID` (set for static pipelines)
+- Falls back to `socket.gethostname()` for dynamic pipelines
+
+---
+
+## Reference Implementations
+
+Study these implementations when building your own orchestrator:
+
+### Kubernetes (recommended starting point)
+**File:** `src/zenml/integrations/kubernetes/orchestrators/kubernetes_orchestrator.py:961`
+
+```python
+def get_orchestrator_run_id(self) -> str:
+    try:
+        return os.environ[ENV_ZENML_KUBERNETES_RUN_ID]
+    except KeyError:
+        # Dynamic pipeline: use pod name
+        return socket.gethostname()
+```
+
+### SageMaker (complex example)
+**File:** `src/zenml/integrations/aws/orchestrators/sagemaker_orchestrator.py:258`
+
+```python
+def get_orchestrator_run_id(self) -> str:
+    # Check multiple environment variables
+    for env in [ENV_ZENML_SAGEMAKER_RUN_ID, "TRAINING_JOB_ARN"]:
+        if env in os.environ:
+            return os.environ[env]
+    
+    # Fall back to processing job config file
+    config_file_path = "/opt/ml/config/processingjobconfig.json"
+    if os.path.exists(config_file_path):
+        with open(config_file_path, "r") as f:
+            # Read job name from config...
+```
+
+### Vertex AI
+**File:** `src/zenml/integrations/gcp/orchestrators/vertex_orchestrator.py:785`
+
+```python
+def get_orchestrator_run_id(self) -> str:
+    try:
+        return os.environ[ENV_ZENML_VERTEX_RUN_ID]
+    except KeyError:
+        raise RuntimeError(
+            f"Unable to read run id from environment variable {ENV_ZENML_VERTEX_RUN_ID}."
+        )
+```
+
+---
+
+## Implementation Checklist
+
+When implementing a new orchestrator:
+
+- [ ] Inherit from `ContainerizedOrchestrator` if your orchestrator runs steps in containers
+- [ ] Implement `get_orchestrator_run_id()` following the static/dynamic patterns above
+- [ ] Implement `submit_pipeline()` for static pipelines
+- [ ] Optionally implement `submit_dynamic_pipeline()` for dynamic pipeline support
+- [ ] Handle `deployment.schedule` if your backend supports scheduling
```

Comment:
`snapshot.schedule`. Also, one could add that the resource settings should also be handled appropriately